Bluetooth: Mesh: Remove msg_cache_next in rx ctx

LingaoM opened the pull request:
Since in bt_mesh_net_recv will call msg_cache_add
and if lpn reject this message, should remove this cache.

But this function sequentially and will not be interrupt by any thread so, we no need a copy var to save this value.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Remove msg_cache_next in rx ctx

Since in bt_mesh_net_recv will call msg_cache_add
and if lpn reject this message, should remove this cache.

But this function sequentially and will not be interrupt
by any thread so, we no need a copy var to save this value.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Fix missing clear dup cache

When lpn reject message, should clear dup cache to
avoid dup check next mesh message.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Balaklaka commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -108,36 +110,45 @@ K_MEM_SLAB_DEFINE(loopback_buf_pool,
-	dup_cache_next %= ARRAY_SIZE(dup_cache);
+	dup.idx	= idx;
+	dup.next %= ARRAY_SIZE(dup_cache[0]);
+	dup_cache[idx][dup.next++] = val;
```
```suggestion
	dup_cache[idx][dup.next] = val;
	dup.next++;
```
Little bit easier to read, and more similar to line 164

Balaklaka commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -108,36 +110,46 @@ K_MEM_SLAB_DEFINE(loopback_buf_pool,
+	dup.idx	= idx;
+	dup.next[dup.idx] %= ARRAY_SIZE(dup_cache[0]);
+	dup_cache[idx][dup.next[dup.idx]] = val;
+	dup.next[dup.idx]++;
```
Why not use [idx] in all off them?

LingaoM replied:
a little bit...

LingaoM commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -108,36 +110,46 @@ K_MEM_SLAB_DEFINE(loopback_buf_pool,
 
 	val = sys_get_be32(tail - 4) ^ sys_get_be32(tail - 8);
 
-	for (i = 0; i < ARRAY_SIZE(dup_cache); i++) {
-		if (dup_cache[i] == val) {
+	idx = val % ARRAY_SIZE(dup_cache[0]);
```
will test and update

LingaoM replied:
I pondered it carefully. It has nothing to do with testing. Your concern is right. This method is more likely to invalidate large-scale caches at the same time, and the additional decryption operations that it brings are compared to the previous iterative query. Not an order of magnitude.

LingaoM replied:
From this, a reverse-order iteration method is proposed, which I think may be better.

LingaoM replied:
I’m do some test, with one node as publish send with generic onoff get mesh message(dst=0xffff, ttl = 5, net_xmit=(8,20)), other ten nodes actor as generic onoff server model, once receive get mesh message, will send generic onoff status(ttl = 5, dst=get message src)


**BEFORE:**
![UdaW76NLlT](https://user-images.githubusercontent.com/26378606/195247700-dc39f2f3-3a76-4c3f-bd58-bbe8f4807e6b.jpg)

**AFTER:**

LingaoM replied:
Although this is a specific scenario, I think it is enough.

PavelVPV replied:
You also need to consider a segmented message (preferably the biggest one) because 1) they will are not decrypted one by one, but once the whole message is received, and 2) the already received segment triggers the reassembly logic at the transport layer.

PavelVPV replied:
But generally yes, the reverse order seems more reasonable.

LingaoM replied:
I don't quite understand why the influence of the transport layer should be considered. This is something of the network layer. It does not depend on the transport layer, because each message is encrypted with a different network nonce. I think the current test can explain the situation.

LingaoM committed:
Bluetooth: Mesh: net cache uses reverse order traversal

Since the net cache is accumulated in positive order, the net
cache uses reverse order traversal, which can filter duplicate
messages more effectively.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -119,14 +119,20 @@ static bool check_dup(struct net_buf_simple *data)
 
-	for (i = 0; i < ARRAY_SIZE(dup_cache); i++) {
-		if (dup_cache[i] == val) {
+	for (i = dup_cache_next; i > 0;) {
+		if (dup_cache[--i] == val) {
+			return true;
+		}
+	}
+
+	for (i = ARRAY_SIZE(dup_cache); i > dup_cache_next;) {
+		if (dup_cache[--i] == val) {
```
With this approach we will go through empty entries until `dup_cache_next` wraps around.

LingaoM replied:
The same is true for the original implementation.

LingaoM replied:
The original intention of this commit is to turn sequential iteration into reverse-order iteration.

LingaoM replied:
In fact, in a network, a cache will be filled soon, and I think it is meaningless to check whether it is blank.

PavelVPV commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -135,8 +141,15 @@ static bool msg_cache_match(struct net_buf_simple *pdu)
 
-	for (i = 0U; i < ARRAY_SIZE(msg_cache); i++) {
-		if (msg_cache[i].src == SRC(pdu->data) &&
+	for (i = msg_cache_next; i > 0U;) {
+		if (msg_cache[--i].src == SRC(pdu->data) &&
+		    msg_cache[i].seq == (SEQ(pdu->data) & BIT_MASK(17))) {
+			return true;
+		}
+	}
+
+	for (i = ARRAY_SIZE(msg_cache); i > msg_cache_next;) {
+		if (msg_cache[--i].src == SRC(pdu->data) &&
```
Same as for `dup_cache_next`. With this approach we will go through empty entries until `msg_cache_next` wraps around.
